pipettes/firmware/can_task.cpp

```diff
@@ -1,82 +1,74 @@
-// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-static FDCAN_RxHeaderTypeDef RxHeader;
-// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-static FDCAN_HandleTypeDef can;
-static constexpr auto buffsize = 64;
-// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-static auto buff = std::array<uint8_t, buffsize>{};
+static auto can_bus_1 = HalCanBus(&fdcan1);
+static auto message_writer_1 = MessageWriter(can_bus_1);
 
-static void run(void *parameter) {
-    parameter = nullptr;
+struct MotorHandler {
+    using MessageType =
+        std::variant<std::monostate, SetSpeedRequest, GetSpeedRequest,
+                     StopRequest, GetStatusRequest, MoveRequest>;
 
-    if (MX_FDCAN1_Init(&can) != HAL_OK) {
-        Error_Handler();
+    void handle(MessageType &m) {
+        std::visit([this](auto o) { this->visit(o); }, m);
     }
 
-    if (HAL_FDCAN_Start(&can) != HAL_OK) {
-        Error_Handler();
-    }
+    void visit(std::monostate &m) {}
+
+    void visit(SetSpeedRequest &m) { message_writer_1.write(NodeId::host, m); }
+
+    void visit(GetSpeedRequest &m) { message_writer_1.write(NodeId::host, m); }
+
+    void visit(StopRequest &m) { message_writer_1.write(NodeId::gantry, m); }
+
+    void visit(GetStatusRequest &m) { message_writer_1.write(NodeId::host, m); }
+
+    void visit(MoveRequest &m) { message_writer_1.write(NodeId::host, m); }
+};
+
+/** The parsed message handler */
+static auto motor_handler = MotorHandler{};
+
+/** The connection between the motor handler and message buffer */
+static auto motor_dispatch_target = FreeRTOSCanDispatcherTarget<
+    256, MotorHandler, can_messages::SetSpeedRequest,
+    can_messages::GetSpeedRequest, can_messages::StopRequest,
+    can_messages::GetStatusRequest, can_messages::MoveRequest>{motor_handler};
 
-    TxHeader.IdType = FDCAN_EXTENDED_ID;
-    TxHeader.TxFrameType = FDCAN_DATA_FRAME;
-    TxHeader.DataLength = FDCAN_DLC_BYTES_8;
-    TxHeader.ErrorStateIndicator = FDCAN_ESI_PASSIVE;
-    TxHeader.BitRateSwitch = FDCAN_BRS_OFF;
-    TxHeader.FDFormat = FDCAN_CLASSIC_CAN;
-    TxHeader.TxEventFifoControl = FDCAN_NO_TX_EVENTS;
-    TxHeader.MessageMarker = 0;
-
-    for (;;) {
-        if (HAL_FDCAN_GetRxFifoFillLevel(&can, FDCAN_RX_FIFO0) > 0) {
-            // Send CAN RX data over debug UART
-            HAL_FDCAN_GetRxMessage(&can, FDCAN_RX_FIFO0, &RxHeader,
-                                   buff.data());
-
-            TxHeader.Identifier = RxHeader.Identifier;
-            TxHeader.DataLength = RxHeader.DataLength;
-            TxHeader.FDFormat = RxHeader.FDFormat;
-
-            if (HAL_FDCAN_AddMessageToTxFifoQ(&can, &TxHeader, buff.data()) !=
-                HAL_OK) {
-                // Transmission request Error
-                Error_Handler();
-            }
+/** Dispatcher to the various handlers */
+static auto dispatcher = Dispatcher(&motor_dispatch_target.parse_target);
```

Comment:
Notice that the Dispatcher will write directly to the parse target and bypass the message buffer target. 

That means that the Task that polls the ROOT CAN message buffer will handle the motor messages.

The idea is that `DispatchParseTarget` and `DispatchMessageBufferTarget` have the same interface. We can either chain a `DispatchMessageBufferTarget` into a `DispatchParseTarget` which requires a task to poll the MessageBuffer or go directly to a `DispatchParseTarget`

I'll make a picture.